File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/render/JumpCircle.java
package net.ccbluex.liquidbounce.features.module.modules.render;

import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.Render3DEvent;
import net.ccbluex.liquidbounce.event.UpdateEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.features.value.FloatValue;
import net.ccbluex.liquidbounce.features.value.IntegerValue;
import org.lwjgl.opengl.GL11;

import java.util.concurrent.CopyOnWriteArrayList;

@ModuleInfo(name = "JumpCircle", category = ModuleCategory.RENDER)
public class JumpCircle extends Module {
    private final IntegerValue redValue = new IntegerValue("Red", 255, 0, 255);
    private final IntegerValue greenValue = new IntegerValue("Green", 255, 0, 255);
    private final IntegerValue blueValue = new IntegerValue("Blue", 255, 0, 255);
    private final IntegerValue radiusValue = new IntegerValue("Radius", 3, 1, 5);
    private final FloatValue widthValue = new FloatValue("Width", 0.5F, 0.1F, 50F);
    private final FloatValue strengthValue = new FloatValue("Strength", 0.02F, 0.01F, 0.2F);
    private final CopyOnWriteArrayList<Circle> circles = new CopyOnWriteArrayList<>();
    private boolean lastOnGround;

    @Override
    public void onEnable(){
        lastOnGround = true;
    }

    @EventTarget
    public void onUpdate(UpdateEvent ignored) {
        if (mc.thePlayer.onGround && !lastOnGround) {
            circles.add(new Circle(mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ, mc.thePlayer.lastTickPosX, mc.thePlayer.lastTickPosY, mc.thePlayer.lastTickPosZ, widthValue.get()));
        }
        lastOnGround = mc.thePlayer.onGround;
    }

    @EventTarget
    public void onRender3D(Render3DEvent ignored) {
        if (!circles.isEmpty()) {
            for (Circle circle : circles) {
                if (circle.add(strengthValue.get()) > radiusValue.get()) {
                    circles.remove(circle);
                    continue;
                }
                GL11.glPushMatrix();
                GL11.glTranslated(
                        circle.posX - mc.getRenderManager().renderPosX,
                        circle.posY - mc.getRenderManager().renderPosY,
                        circle.posZ - mc.getRenderManager().renderPosZ
                );
                GL11.glEnable(GL11.GL_BLEND);
                GL11.glEnable(GL11.GL_LINE_SMOOTH);
                GL11.glDisable(GL11.GL_TEXTURE_2D);
                GL11.glDisable(GL11.GL_DEPTH_TEST);
                GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);

                GL11.glLineWidth(circle.width);
                GL11.glColor4f(redValue.get() / 255.0F, greenValue.get() / 255.0F, blueValue.get() / 255.0F, (radiusValue.get() - circle.radius) / radiusValue.get());
                GL11.glRotatef(90F, 1F, 0F, 0F);

                GL11.glBegin(GL11.GL_LINE_STRIP);

                for (int i = 0; i <= 360; i += 5) { // You can change circle accuracy  (60 - accuracy)
                    GL11.glVertex2f(
                            (float) (Math.cos(i * Math.PI / 180.0) * circle.radius),
                            (float) (Math.sin(i * Math.PI / 180.0) * circle.radius)
                    );
                }

                GL11.glEnd();

                GL11.glDisable(GL11.GL_BLEND);
                GL11.glEnable(GL11.GL_TEXTURE_2D);
                GL11.glEnable(GL11.GL_DEPTH_TEST);
                GL11.glDisable(GL11.GL_LINE_SMOOTH);
                GL11.glPopMatrix();
            }
        }
    }

    static class Circle {
        public double posX, posY, posZ, lastTickPosX, lastTickPosY, lastTickPosZ;
        public float radius, width;
        Circle(double posX, double posY, double posZ, double lastTickPosX, double lastTickPosY, double lastTickPosZ, float width) {
            this.posX = posX;
            this.posY = posY;
            this.posZ = posZ;
            this.lastTickPosX = lastTickPosX;
            this.lastTickPosY = lastTickPosY;
            this.lastTickPosZ = lastTickPosZ;
            this.width = width;
        }
        public double add(double radius) {
            this.radius += radius;
            return this.radius;
        }
    }
}

File: src/main/java/net/ccbluex/liquidbounce/features/module/modules/render/JumpCircle.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.features.module.modules.render

import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.event.EventTarget
import net.ccbluex.liquidbounce.event.Render3DEvent
import net.ccbluex.liquidbounce.event.UpdateEvent
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.features.value.BoolValue
import net.ccbluex.liquidbounce.features.value.FloatValue
import net.ccbluex.liquidbounce.features.value.IntegerValue
import net.ccbluex.liquidbounce.utils.MathUtils.toRadians
import net.ccbluex.liquidbounce.utils.render.ColorUtils
import net.ccbluex.liquidbounce.utils.render.RenderUtils
import org.lwjgl.opengl.GL11
import java.awt.Color
import kotlin.math.cos
import kotlin.math.sin

@ModuleInfo(name = "JumpCircle", category = ModuleCategory.RENDER)
class JumpCircle : Module() {

    val disappearTime = IntegerValue("Time", 1000, 1000,3000)
    val radius = FloatValue("Radius", 2f, 1f,5f)

    val rainbow = BoolValue("Rainbow", false)

    val start = FloatValue("Start", 0.5f, 0f,1f)
    val end = FloatValue("End", 0.3f, 0f,1f)

    val circles = mutableListOf<Circle>()
    var lastOnGround = false

    @EventTarget
    fun onUpdate(event: UpdateEvent) {
        if (!mc.thePlayer.onGround && lastOnGround) {
            circles.add(Circle(System.currentTimeMillis(), mc.thePlayer.posX, mc.thePlayer.posY, mc.thePlayer.posZ))
            lastOnGround = false
        }
        if (mc.thePlayer.onGround) {
            lastOnGround = true
        }
    }

    @EventTarget
    fun onRender3D(event: Render3DEvent) {
        circles.removeIf { System.currentTimeMillis() > it.time + disappearTime.get() }

        GL11.glPushMatrix()

        GL11.glEnable(GL11.GL_BLEND)
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA)
        GL11.glDisable(GL11.GL_CULL_FACE)
        GL11.glDisable(GL11.GL_TEXTURE_2D)
        GL11.glDisable(GL11.GL_DEPTH_TEST)
        GL11.glDepthMask(false)
        GL11.glDisable(GL11.GL_ALPHA_TEST)
        GL11.glShadeModel(GL11.GL_SMOOTH)

        circles.forEach { it.draw() }

        GL11.glDisable(GL11.GL_BLEND)
        GL11.glEnable(GL11.GL_CULL_FACE)
        GL11.glEnable(GL11.GL_TEXTURE_2D)
        GL11.glEnable(GL11.GL_DEPTH_TEST)
        GL11.glDepthMask(true)
        GL11.glEnable(GL11.GL_ALPHA_TEST)
        GL11.glShadeModel(GL11.GL_FLAT)

        GL11.glPopMatrix()
    }

    class Circle(val time: Long, val x: Double, val y: Double, val z: Double){
        val jumpModule = LiquidBounce.moduleManager.getModule(JumpCircle::class.java) as JumpCircle

        fun draw() {

            val dif = (System.currentTimeMillis() - time)
            val c = 255 - (dif / jumpModule.disappearTime.get().toFloat()) * 255

            GL11.glPushMatrix()

            GL11.glTranslated(
                x - mc.renderManager.viewerPosX,
                y - mc.renderManager.viewerPosY,
                z - mc.renderManager.viewerPosZ
            )

            GL11.glBegin(GL11.GL_TRIANGLE_STRIP)
            for (i in 0..360) {
                val color = if (jumpModule.rainbow.get()) Color.getHSBColor(i / 360f, 1f, 1f)
                else ColorUtils.hsbTransition(jumpModule.start.get(), jumpModule.end.get(), i)

                val x = (dif * jumpModule.radius.get() * 0.001 * sin(i.toDouble().toRadians()))
                val z = (dif * jumpModule.radius.get() * 0.001 * cos(i.toDouble().toRadians()))

                RenderUtils.glColor(color.red, color.green, color.blue, 0)
                GL11.glVertex3d(x / 2, 0.0, z / 2)

                RenderUtils.glColor(color.red, color.green, color.blue, c.toInt())
                GL11.glVertex3d(x, 0.0, z)
            }
            GL11.glEnd()

            GL11.glPopMatrix()
        }
    }

}
